### PR TITLE
[Feature] Add RandomForestClassifier to linfa-trees

### PR DESCRIPTION
**Pull request by maxprogrammer007**
This PR extends the `linfa-trees` crate by introducing a new **Random Forest** classifier. It builds upon the existing Decision Tree implementation to provide an ensemble method that typically outperforms a single tree by training many trees on bootstrapped subsets of both **rows** and **columns** (features), and then aggregating their predictions via majority voting.

---

### 🚀 What’s Added

1. **`src/decision_trees/random_forest.rs`**

   * **`RandomForestParams` / `RandomForestValidParams`**: hyperparameters (`n_trees`, `max_depth`, `feature_subsample`, `seed`) with validation via `ParamGuard`.
   * **`RandomForestClassifier`**: stores a `Vec<DecisionTree>` and the per-tree feature indices used for training.
   * **`Fit` implementation**:

     * Bootstrap rows.
     * Randomly subsample features per tree.
     * Train each `DecisionTree` on its slice.
   * **`Predict` implementation**:

     * For each tree, select the same feature slice on the test data.
     * Invoke `tree.predict(&sub_x)` (returns `Array1<usize>`).
     * Accumulate votes and return the argmax for each sample.

2. **Exports**

   * Updated `src/decision_trees/mod.rs` and `src/lib.rs` to re-export `RandomForestParams` and `RandomForestClassifier`.

3. **Example**

   * `examples/iris_random_forest.rs`: demonstrates loading the Iris dataset, training a Random Forest, printing the confusion matrix and accuracy.

4. **Unit Test**

   * `tests/random_forest.rs`: an integration test asserting ≥ 90 % accuracy on Iris with fixed RNG seed for reproducibility.

5. **Dependencies**

   * Added `rand = "0.8"` to `linfa-trees/Cargo.toml` for RNG and sampling utilities.

6. **README**

   * Extended `README.md` with a “Random Forest Classifier” section, usage example, and run instructions.

---

### 🧐 Motivation

* **Ensemble performance**: Random Forests often reduce variance and improve generalization compared to a single decision tree.
* **Feature importance**: Subsampling features per tree provides insight into feature usefulness.
* **API consistency**: Follows Linfa’s `Fit` / `Predict` / `ParamGuard` conventions and integrates cleanly with `Dataset`.

---

### 🔍 Files Changed

```
algorithms/linfa-trees/
├─ Cargo.toml            # + rand = "0.8"
├─ src/
│  ├─ decision_trees/
│  │  ├─ algorithm.rs    # no change
│  │  ├─ mod.rs          # + pub mod random_forest;
│  │  └─ random_forest.rs # NEW
│  └─ lib.rs             # + pub use decision_trees::random_forest::{…}
├─ examples/
│  └─ iris_random_forest.rs # NEW
└─ tests/
   └─ random_forest.rs   # NEW
```

---

### 📦 Example

```bash
cargo run --release --example iris_random_forest
```

```text
classes    | 0  | 1  | 2
--------------------------------
0          | 50 |  0 |  0
1          |  0 | 48 |  2
2          |  0 |  1 | 49

Accuracy: 0.97
```

---

### ✅ Checklist

* [x] Implements `ParamGuard` for hyperparameter validation
* [x] Implements `Fit<Array2<F>, Array1<usize>>`
* [x] Implements `Predict<Array2<F>, Array1<usize>>` with correct feature‐slice logic
* [x] Example runs without errors (`cargo run --example iris_random_forest`)
* [x] Unit test passes (`cargo test`)
* [x] README updated with usage snippet
* [x] `rand` dependency added

---

Thank you for reviewing! I’m happy to address any feedback or suggestions.

### CODE REVIEW

**Comment by relf** [comment]
@maxprogrammer007, I've just merged #392 which introduces `linfa-ensemble` and bagging algorithm. 
This mirrors a bit (though far from being as complete) the [scikit-learn structure](https://scikit-learn.org/stable/modules/ensemble.html#ensemble).
To get proper `RandomForest` algorithm in this new `ensemble` sub-crate we need to add features sub-sampling. 

So if you agree, I suggest you could reuse part of your code to implement a `RandomForest` in `linfa-ensemble` based on `EnsembleLearner` and get something like:
```rust
struct RandomForest<F: Float, L: Label> {
    ensemble_learner: EnsembleLearner<DecisionTree<F, L>>,
    bootstrap_features_ratio: f64,
    feature_indices: Vec<Vec<usize>>
}
```  
A step further, would be to manage feature subsampling directly in `DecisionTree` then `RandomForest<F, L>` would be just a thin wrapper around `EnsembleLearner<DecisionTree<F, L>>`.
If you proceed maybe start over with a new PR and close this one. What do you think?

**Comment by maxprogrammer007** [comment]
@relf sure i will proceed with new PR.
